### PR TITLE
Move to std::future and add support for async/await

### PR DESCRIPTION
**Pull request by jonhoo**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -21,20 +21,18 @@ log = ["tracing/log"]
 default = ["tracing"]
 
 [dependencies]
-tower-service = "0.2"
+tower-service = "0.3.0-alpha.1"
```
Comment (LucioFranco) [inline Cargo.toml]: I've been using `=` just to ensure everything lines up.

File: Cargo.toml [hunk 2]
```diff
@@ -21,20 +21,18 @@ log = ["tracing/log"]
...
 tower-load = { git = "https://github.com/tower-rs/tower.git" }
-futures = "0.1"
-tokio-executor = "0.1.5"
-tokio-sync = "0.1"
+futures-preview = "0.3.0-alpha.18"
```
Comment (LucioFranco) [inline Cargo.toml]: you most likely want to depend on `futures-util` and `futures-core` intead of the whole facade.

File: src/multiplex/client.rs [hunk 3]
```diff
@@ -211,124 +186,82 @@ where
...
-                            tracing::event!(Level::TRACE, "Sink full; queueing");
-                            drop(guard);
-                        }
-                        #[cfg(not(feature = "tracing"))]
-                        let span = ();
-                        self.waiting = Some((id, ClientRequest { req, span, res }));
-                    } else {
+                        .map_err(Error::from_sink_error)?;
+                    tracing::event!(Level::TRACE, "request sent");
+                    drop(guard);
```
Comment (LucioFranco) [inline src/multiplex/client.rs]: will this not fail without the tracing feature?

File: src/multiplex/client.rs [hunk 4]
```diff
@@ -22,10 +24,10 @@ pub trait TagStore<Request, Response> {
     type Tag: Eq;
 
     /// Assign a fresh tag to the given `Request`, and return that tag.
-    fn assign_tag(&mut self, r: &mut Request) -> Self::Tag;
+    fn assign_tag(self: Pin<&mut Self>, r: &mut Request) -> Self::Tag;
```
Comment (LucioFranco) [inline src/multiplex/client.rs]: why does this need to be pin?

Comment (jonhoo) [reply src/multiplex/client.rs]: Because we give out `&mut` to it from behind a `Pin`. The alternative would be to require `TagStore + Unpin` everywhere.

Comment (jonhoo) [reply src/multiplex/client.rs]: Specifically, the issue is that the same `T` that is `:TagStore` is _also_ our `Stream + Sink`, so we are going to be giving out `Pin<&mut T>`, which means we can't then hand out `&mut T`.

Comment (jonhoo) [reply src/multiplex/client.rs]: Elaborated on this in the docs for `TagStore` now.

**Pull request merged**